File: cpy0/c/call.c
#include "call.h"

#include "eval.h"
#include "util.h"

PyValue *py_call(PyRuntime *rt, PyValue *callable, int argc, PyValue **argv) {
    if (callable->type == PY_BUILTIN_FUNCTION) {
        return callable->as.cfunc->fn(rt, argc, argv);
    }
    if (callable->type == PY_FUNCTION) {
        PyFunction *fn = callable->as.func;
        if (argc != fn->param_count) {
            die("%s() expected %d arguments, got %d", fn->name, fn->param_count, argc);
        }
        PyEnv *local = env_new(fn->closure);
        for (int i = 0; i < argc; i++) {
            env_set(local, fn->params[i], argv[i]);
        }
        ExecResult result = exec_function_body(rt, local, fn->body);
        if (result.has_return) {
            return result.value;
        }
        return py_none();
    }
    die("object is not callable");
    return py_none();
}

File: _bak/cpy0/rust/src/call.rs
use crate::eval::exec_function_body;
use crate::value::{PyEnv, PyRuntime, PyValue};

pub fn py_call(rt: &mut PyRuntime, callable: &PyValue, argc: usize, argv: &mut [PyValue]) -> PyValue {
    match callable.py_type {
        crate::value::PyType::BuiltinFunction => {
            let fn_ = callable.as_cfunc.as_ref().unwrap();
            (fn_.fn_)(rt, argc, argv)
        }
        crate::value::PyType::Function => {
            let fn_ = callable.as_func.as_ref().unwrap();
            if argc != fn_.param_count {
                crate::util::die(&format!(
                    "{}() expected {} arguments, got {}",
                    fn_.name, fn_.param_count, argc
                ));
            }

            let closure_ptr = fn_.closure;
            let parent_env = unsafe { &*closure_ptr };
            let mut local = PyEnv::env_new(Some(parent_env.clone()));
            for i in 0..argc {
                local.set(&fn_.params[i], argv[i].clone());
            }

            let result = exec_function_body(rt, &mut local, &fn_.body);
            if result.has_return {
                if let Some(v) = result.value {
                    return v;
                }
            }
            PyValue::new_none()
        }
        _ => {
            crate::util::die("object is not callable");
            PyValue::new_none()
        }
    }
}